File: frontend/src/snake/ServerMessage.ts
export type Direction = { type: "Up" | "Down" | "Left" | "Right" };

export type PowerUps = {
	type: "ExtraLife"
} | {
	type: "AddLength"
} | {
	type: "ShrinkOpponent",
	opponent: string
} | {
	type: "FreezeOpponent",
	opponent: string
} | {
	type: "Revive"
}

export type SentSnake = {
	type: "Alive",
	user_id: string,
	head: [number, number],
	head_direction: Direction,
	blocks: [Direction, number][],
	invulnerable: boolean,
	frozen: boolean,
	has_extra_life: boolean
} | {
	type: "Dead",
	user_id: string,
	revive_left: number
}

export type ReadyStatus = {
	user_id: string,
	ready: boolean
}

export type AmountSpent = {
	user_id: string,
	amount_spent: number
}

export type RecentPowerUp = {
	user_id: string,
	power_up: PowerUps
}

export type ServerMessage  ={
	type: "Authenticated"
} | {
	type: "ReadyStatus",
	status: ReadyStatus[]
} | {
	type: "StartGame"
} | {
	type: "GameOver",
	winner: string,
	amounts_spent: AmountSpent[]
} | {
	type: "GameState",
	apples: [number, number][],
	snakes: SentSnake[],
	just_ate_apple: string[],
	recent_power_ups: RecentPowerUp[]
}


export type ClientMessage = {
	type: "Authenticate",
	access_token: string,
	game_id: string
} | {
	type: "UsePowerUp",
	power_up: PowerUps
} | {
	type: "SetDirection",
	direction: Direction
} | {
	type: "SetReady",
	ready: boolean
};

File: multiplayer/src/games_server/server_message.rs
use serde::Serialize;
use crate::games_server::all_games_state::game_state::{Direction, Snake};
use crate::games_server::power_ups::PowerUps;

#[derive(Debug, Serialize)]
#[serde(tag = "type")]
#[derive(Clone)]
pub enum SentSnake {
    Alive { user_id: String, head: (u32, u32), head_direction: Direction, blocks: Vec<(Direction, u32)>, invulnerable: bool, frozen: bool, has_extra_life: bool },
    Dead { user_id: String, revive_left: u64 },
}

impl Into<SentSnake> for &Snake {
    fn into(self) -> SentSnake {
        match self {
            Snake::Alive(snake) => SentSnake::Alive {
                invulnerable: snake.invulnerable_for.is_some(),
                frozen: snake.frozen_for.is_some(),
                user_id: snake.user_id.to_string(),
                head: snake.head,
                head_direction: snake.head_direction.clone(),
                blocks: snake.blocks.clone(),
                has_extra_life: snake.has_extra_life,
            },
            Snake::Dead { user_id, ticks_to_revive, .. } => SentSnake::Dead { 
                user_id: user_id.clone(),
                revive_left: match ticks_to_revive {
                    Some(ticks) => *ticks,
                    None => 0,
                }
            }
        }
    }
}

#[derive(Debug, Serialize)]
#[derive(Clone)]
pub struct ReadyStatus {
    pub user_id: String,
    pub ready: bool
}

#[derive(Debug, Serialize)]
#[derive(Clone)]
pub struct AmountSpent {
    pub user_id: String,
    pub amount_spent: f64
}

#[derive(Debug, Serialize)]
#[derive(Clone)]
pub struct RecentPowerUp {
    pub user_id: String,
    pub power_up: PowerUps
}

#[derive(Debug, Serialize, Clone)]
#[serde(tag = "type")]
pub enum ServerMessage {
    Authenticated,
    /// vector of 
    ReadyStatus{status:Vec<ReadyStatus>},
    StartGame,
    GameOver {winner: String, amounts_spent: Vec<AmountSpent>},
    GameState { apples: Vec<(u32, u32)>, snakes: Vec<SentSnake>, just_ate_apple: Vec<String>, recent_power_ups: Vec<RecentPowerUp> }
}
